quarkus/runtime/src/main/java/org/keycloak/provider/quarkus/QuarkusClientConnectionFilter.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.provider.quarkus;
+
+import io.quarkus.vertx.web.RouteFilter;
+import io.vertx.core.Handler;
+import io.vertx.core.http.HttpServerRequest;
+import io.vertx.ext.web.RoutingContext;
+import org.keycloak.common.ClientConnection;
+import org.keycloak.common.util.Resteasy;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.services.filters.AbstractClientConnectionServletFilter;
+
+import javax.inject.Inject;
+import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.container.PreMatching;
+import javax.ws.rs.ext.Provider;
+
+@PreMatching
+@Provider
+public class QuarkusClientConnectionFilter implements javax.ws.rs.container.ContainerRequestFilter {
+
+    @Inject
+    RoutingContext routingContext;
+
+    @Override
+    public void filter(ContainerRequestContext containerRequestContext) {
+        HttpServerRequest request = routingContext.request();
+        ClientConnection connection = createConnection(request);
+
+        AbstractClientConnectionServletFilter servletFilter = new AbstractClientConnectionServletFilter() {
+            @Override
+            protected void nextFilter(KeycloakSession session) {
+                Resteasy.pushContext(KeycloakSession.class, session);
+                Resteasy.pushContext(ClientConnection.class, connection);
+
+                routingContext.put("KeycloakSession", session);
+            }
+        };
+        routingContext.put("AbstractClientConnectionServletFilter", servletFilter);
+        servletFilter.filter();
+    }
+
+    @RouteFilter()
+    public void vertxFilter(RoutingContext rc) {
```

Comment:
I remember using filter differently where we basically dispatch the request from the eventloop so that we can control all the request lifecycle.

I think it should work better as well as serve as an important point for any other change er might need that affects the thread execution.